test/util/testnode/utils.go

```diff
@@ -108,14 +118,27 @@ func GetFreePort() (int, error) {
 	return l.Addr().(*net.TCPAddr).Port, nil
 }
 
-// mustGetFreePort returns a free port. Panics if no free ports are available or
-// an error is encountered.
-func mustGetFreePort() int {
-	port, err := GetFreePort()
+// isPortAvailable checks if a port is available by attempting to listen on it.
+func isPortAvailable(port int) bool {
+	l, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
 	if err != nil {
-		panic(err)
+		return false
+	}
+	defer l.Close()
+	return true
+}
+
+// GetDeterministicPort returns a deterministic port using an atomic counter.
+// This eliminates race conditions by ensuring each call gets a unique port.
+// It checks port availability and increments until it finds an open port.
+func GetDeterministicPort() int {
+	for {
+		port := int(portCounter.Add(1))
+		if isPortAvailable(port) {
+			return port
+		}
+		// If port is not available, the loop will continue with the next increment
```

Comment:
iirc we tried something like this before @rach-id ? if so, do you remember if there were other problems with it?